src/System.Diagnostics.StackTrace.Symbols/src/System/Diagnostics/StackTrace/Symbols.cs

```diff
@@ -0,0 +1,222 @@
+                                                j--;
+                                            }
+                                            break;
+                                        }
+                                    }
+
+                                    // If we didn't find a match, default to the last sequence point
+                                    if (j == sequencePointCount) {
+                                        j--;
+                                    }
+
+                                    while (lines[j] == SequencePoint.HiddenLine && j > 0) {
+                                        j--;
+                                    }
+
+                                    if (lines[j] != SequencePoint.HiddenLine) {
+                                        sourceLine = lines[j];
+                                        sourceColumn = columns[j];
+                                    }
+                                    sourceFile = reader.GetString(reader.GetDocument(documents[j]).Name);
+                                }
+                            }
+                            catch {
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Returns the portable PDB reader for the assembly path
+        /// </summary>
+        /// <param name="assemblyPath">file path of the assembly or null</param>
+        /// <param name="inMemorySymbols">in memory PDB address or zero</param>
+        /// <param name="inMemorySymbolsSize">in memory PDB size</param>
+        /// <param name="reader">returns the reader</param>
+        /// <returns>provider</returns>
+        static IDisposable GetReader(string assemblyPath, IntPtr inMemorySymbols, int inMemorySymbolsSize, out MetadataReader reader)
+        {
+            reader = null;
+            MetadataReaderProvider provider = null;
+
+            if (assemblyPath != null) {
+                try {
+                    uint stamp;
+                    int age;
+                    Guid guid;
+
+                    string pdbName = GetPdbPathFromPeStream(assemblyPath, out age, out guid, out stamp);
+                    if (pdbName != null) {
+                        Stream pdbStream = File.OpenRead(pdbName);
+
+                        // Need to always return provider so that it will be disposed
+                        provider = MetadataReaderProvider.FromPortablePdbStream(pdbStream);
+                        MetadataReader rdr = provider.GetMetadataReader();
+
+                        // Validate that the PDB matches the assembly version
+                        if (age == 1) {
+                            if (IdEquals(rdr.DebugMetadataHeader.Id, guid, stamp)) {
+                                reader = rdr;
+                            }
+                        }
+                    }
+                }
+                catch {
```

Comment:
I propose we come up with a separate strategy for avoiding the endless recursive loop. Identifying every block of code that might throw exceptions and wrapping it in a try/catch seems error-prone. There is also the opportunity for code we don't directly control such as everything inside System.Reflection.Metadata.dll to catch one of its own exceptions and try to evaluate the StackTrace. How about adding a ThreadStatic recursion guard field to the code in mscorlib that tries to populate the stack trace file information? If we detect a recursive call we could avoid populating any file/line/column info for the recursive exception.

Seperately from the recursion guard, if there are anticipated errors ideally I'd do something to avoid throwing an exception at all. If that isn't possible I hope we can catch precise exception types rather than using the scarier catch-and-ignore-everything pattern : )